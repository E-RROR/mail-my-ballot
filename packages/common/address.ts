export interface Address {
  queryAddr: string
  fullAddr: string
  houseNumber: string
  road: string
  unit: string
  city: string
  state: string
  postcode: string
  country: string
  county: string
}

export const uspsAddressTwoLines = (
  {
    houseNumber,
    road,
    unit,
    city,
    state,
    postcode
  }: Address
): [string, string] => {
  return [
    `${houseNumber} ${road}` + (unit ? ` Unit ${unit}` : ''),
    `${city}, ${state} ${postcode}`
  ]
}

export const uspsAddressOneLine = (address: Address): string => {
  const [x, y] = uspsAddressTwoLines(address)
  return x + ', ' + y
}
