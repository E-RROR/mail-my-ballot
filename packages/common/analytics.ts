export interface Analytics {
  facebookId?: string
}
