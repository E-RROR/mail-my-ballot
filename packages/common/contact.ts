import { State } from './states'

interface BaseContact {
  state: State
  official: string
  emails?: string[]
  faxes?: string[]
}

export interface MichiganContact extends BaseContact {
  state: "Michigan"
  city: string
  county: string
  emails: string[]
  faxes: string[]
  phones: string[]
}

export interface FloridaContact extends BaseContact{
  state: "Florida"
  county: string
  emails: string[]
  url: string
}

export type Contact = (
  | MichiganContact
  | FloridaContact
)
