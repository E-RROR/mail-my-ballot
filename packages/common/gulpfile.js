const gulp = require('gulp')
const run = require('@tianhuil/gulp-run-command').default

gulp.task('test', run('jest'))
