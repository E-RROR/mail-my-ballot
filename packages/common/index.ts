export function processEnvOrThrow(key: string): string {
  const val = process.env[key]
  if (!val) throw new Error(`Need to set environment variable ${key}`)
  return val
}

export const isProd = (): boolean => (process.env.NODE_ENV === 'production')

export * from './util'
export * from './address'
export * from './trpc'
export * from './states'
