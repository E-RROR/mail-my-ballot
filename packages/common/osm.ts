import { Address } from "."
import { isState, State } from "./states"

/*
  In theory, this module could exist on the client, but we need it in common for
  testing on the server.  See `contact/index.test.ts`.
*/

const fetch = (typeof window !== 'undefined' && window.fetch) ? window.fetch : require('node-fetch')

const getJson = async <T>(url: string): Promise<T> => {
  const result = await fetch(url)
  return await result.json()
}

interface BasicAddress {
  houseNumber: string
  road: string
  unit: string
  city: string
  country: string
  state: string
  postcode: string
  county: string
}

type GenericRecord = Record<string, unknown>

const osmQuery = async <T>(
  query: string,
  parseSummary: (summary: GenericRecord) => T | null,
  parseDetail?: (detail: GenericRecord) => T | null,
): Promise<T | null> => {
  const summary = (await getJson<Array<GenericRecord>>(`https://nominatim.openstreetmap.org/search/${query}?format=json&countrycodes=us`))[0]
  if (!summary) return null

  const summaryResult = parseSummary(summary)
  if (summaryResult) return summaryResult
  
  if (!parseDetail) return null
  const osmId = (summary.osm_type as string)[0].toUpperCase() + summary.osm_id
  const detail = (await getJson<Array<GenericRecord>>(`https://nominatim.openstreetmap.org/lookup?osm_ids=${osmId}&format=json`))[0]
  if (!detail) return null
  const detailResult = parseDetail(detail)
  return detailResult
}

class PartsParser {
  parts: string[]
  displayName: string

  constructor(displayName: string) {
    this.displayName = displayName
    this.parts = displayName.split(',').map(x => x.trim()).reverse()
  }

  parsePart(re?: RegExp | ((part: string) => boolean)): string | undefined {
    const match = (
      (typeof re === 'undefined') ? (_: string) => true : 
      (typeof re === 'function') ? re :
      (part: string) => !!part.match(re)
    )
    if (this.parts[0] && match(this.parts[0])) {
      return (this.parts.shift() as string)
    }
    return undefined
  }

  parse() {
    const country = this.parsePart(/^United States of America$/)
    const postcode = this.parsePart(/^\d{5}/)
    const state = this.parsePart(isState) as State | null
    const county = this.parsePart(/County$/)
    const city = this.parsePart()

    return {
      country,
      postcode,
      state,
      county,
      city,
    }
  }
}

export const fetchState = async (zip: string): Promise<State | null> => {
  const query = `${zip} United States`
  return osmQuery(query, (summary) => {
    const parser = new PartsParser(summary.display_name as string)
    return parser.parse().state || null
  })
}

const parseDisplayName = (displayName: string, unit: string): BasicAddress | null => {
  /*
  Parse a display_name field from osm sumary result, e.g.
  2125, Butterfield Road, Regents Club Of Troy, Big Beaver, Troy, Oakland County, Michigan, 48084, United States of America
  */

  // 
  const parts = displayName.split(',')
  if(parts.length < 7) return null
  return {
    houseNumber: parts[0].trim(),
    road: parts[1].trim(),
    unit,
    city: parts[parts.length - 5].trim(),
    county: parts[parts.length - 4].trim(),
    state: parts[parts.length - 3].trim(),
    postcode: parts[parts.length - 2].trim(),
    country: parts[parts.length - 1].trim(),
  }
}

export const geocode = async (queryAddr: string, unit = ''): Promise<Address | null> => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const summary = (await getJson<Array<any>>(`https://nominatim.openstreetmap.org/search/${queryAddr}?format=json&countrycodes=us`))[0]
  if (!summary) {
    return null
  }
  const osmId = summary.osm_type[0].toUpperCase() + summary.osm_id
  const summaryAddress = parseDisplayName(summary.display_name, unit)
  if (summaryAddress) {
    return {
      queryAddr,
      fullAddr: summary.display_name,
      ...summaryAddress
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const detail = (await getJson<Array<any>>(`https://nominatim.openstreetmap.org/lookup?osm_ids=${osmId}&format=json`))[0]
  if (!detail) {
    return null
  }
  const { address } = detail
  const {
    house_number: rawHouseNumber,
    road,
    city,
    state,
    postcode,
    country,
    county
  } = address

  // OSM is often missing the house number on response.  So we'll infer it
  const fullAddr = summary.display_name
  const houseNumber = rawHouseNumber ? rawHouseNumber : fullAddr.split(',')[0].trim()

  return {
    queryAddr,
    fullAddr,
    houseNumber,
    road,
    unit,
    city,
    state,
    postcode,
    country,
    county,
  }
}
