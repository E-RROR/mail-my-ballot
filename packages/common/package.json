{
  "name": "@vbm/common",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@tianhuil/simple-trpc": "^1.5.0"
  },
  "devDependencies": {
    "@babel/preset-typescript": "^7.8.3",
    "@types/jest": "^24.0.0",
    "ts-jest": "^24.3.0",
    "typescript": "~3.7.2"
  },
  "scripts": {
    "test": "jest"
  }
}
