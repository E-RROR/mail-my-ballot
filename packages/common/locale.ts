import { State, isState } from './states'
import { Address} from './address'

export interface Locale<S extends State = State> {
  city: string
  county: string
  state: S
}

export const toLocale = (address: Address): Locale | null => {
  const { state, city, county } = address
  if (!isState(state) || !city || !county) return null
  return { state, city, county }
}
