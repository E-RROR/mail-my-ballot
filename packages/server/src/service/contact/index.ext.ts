import { Address, sampleAddresses, toLocale, toContactMethod, AvailableState, isAvailableState, Locale, wait } from '../../common'
import { toContact } from '.'
import { geocode } from '../gm'
import fs from 'fs'

const encoding = 'utf8'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Func<T extends Record<string, any>> = (_: string) => Promise<T>
const cache = <T>(f: Func<T>): Func<T> => {
  return async (arg: string) => {
    const path = `${__dirname}/cache/${arg}.json`
    if (fs.existsSync(path)) {
      const data = fs.readFileSync(path, { encoding } )
      return JSON.parse(data)
    } else {
      const ret = await f(arg)
      const data = JSON.stringify(ret)
      fs.writeFileSync(path, data, { encoding })
    }
  }
}

test('OSM is returning stable results', async (cb) => {
  /*
    Testing geocode is more appropriately with `osm.ts`.  However, it has to be
    here because we want to check that the resulting contact and method objects are not
    null, and this code is (and can only be) available on the server.
  */

  const results = await Promise.all(
    sampleAddresses.map(async (addrData, index) => {
      await wait(index * 0) // space calls out if we're running for firs time
      const result = await cache(geocode)(addrData.address)
      expect(result).toBeTruthy()
      return result as Address
    })
  )

  results.forEach(async (result) => {
    const locale = toLocale(result)
    expect(locale).toBeTruthy()
    expect(isAvailableState((locale as Locale).state)).toBeTruthy()
    const contact = await toContact(locale as Locale<AvailableState>)
    expect(contact).toBeTruthy()
    const method = toContactMethod(contact)
    expect(method).toBeTruthy()
  })

  cb()
})
