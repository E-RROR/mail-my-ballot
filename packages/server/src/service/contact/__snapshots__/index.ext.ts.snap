// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OSM is returning stable results: Berrien County, Michigan 1`] = `
Object {
  "city": "Saint Joseph",
  "country": "United States of America",
  "county": "Berrien County",
  "fullAddr": "700, Broad Street, Saint Joseph, Berrien County, Michigan, 49085, United States of America",
  "houseNumber": "700",
  "postcode": "49085",
  "queryAddr": "700 Broad Street St Joseph, MI 49085",
  "road": "Broad Street",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Bibb County, Georgia 1`] = `
Object {
  "city": "Macon",
  "country": "United States of America",
  "county": "Bibb County",
  "fullAddr": "305, Coliseum Drive, Shirley Hills, North Highlands, Macon, Bibb County, Georgia, 31217, United States of America",
  "houseNumber": "305",
  "postcode": "31217",
  "queryAddr": "305 Coliseum Dr, Macon, GA 31217",
  "road": "Coliseum Drive",
  "state": "Georgia",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Broward County, Florida 1`] = `
Object {
  "city": "Fort Lauderdale",
  "country": "United States of America",
  "county": "Broward County",
  "fullAddr": "2000, West Commercial Boulevard, Fort Lauderdale, Broward County, Florida, 33309, United States of America",
  "houseNumber": "2000",
  "postcode": "33309",
  "queryAddr": "2000 W Commercial Blvd, Fort Lauderdale, FL 33309",
  "road": "West Commercial Boulevard",
  "state": "Florida",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Brown County, Wisconsin 1`] = `
Object {
  "city": "Green Bay",
  "country": "United States of America",
  "county": "Brown County",
  "fullAddr": "300, North Broadway, Fort Howard, Green Bay, Brown County, Wisconsin, 54303, United States of America",
  "houseNumber": "300",
  "postcode": "54303",
  "queryAddr": "300 N Broadway, Green Bay, WI 54303",
  "road": "North Broadway",
  "state": "Wisconsin",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Canton, Wayne County, Michigan 1`] = `
Object {
  "city": "Canton",
  "country": "United States of America",
  "county": "Wayne County",
  "fullAddr": "45525, Hanford Road, Cherry Hill, Canton, Wayne County, Michigan, 48187, United States of America",
  "houseNumber": "45525",
  "postcode": "48187",
  "queryAddr": "45525 Hanford Rd, Canton, MI 48187",
  "road": "Hanford Road",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Dane County, Wisconsin 1`] = `
Object {
  "city": "Madison",
  "country": "United States of America",
  "county": "Dane County",
  "fullAddr": "L'Etoile, South Pinckney Street, Judge Doyle Square, Madison, Dane County, Wisconsin, 53703, United States of America",
  "houseNumber": "L'Etoile",
  "postcode": "53703",
  "queryAddr": "1 S Pinckney St, Madison, WI 53703",
  "road": "South Pinckney Street",
  "state": "Wisconsin",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Dearborn Heights, Wayne County, Michigan 1`] = `
Object {
  "city": "Dearborn Heights",
  "country": "United States of America",
  "county": "Wayne County",
  "fullAddr": "24624, West Warren Street, Dearborn Heights, Wayne County, Michigan, 48127, United States of America",
  "houseNumber": "24624",
  "postcode": "48127",
  "queryAddr": "24624 W Warren St, Dearborn Heights 48127, MI",
  "road": "West Warren Street",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Dearborn, Wayne County, Michigan 1`] = `
Object {
  "city": "Dearborn",
  "country": "United States of America",
  "county": "Wayne County",
  "fullAddr": "22100, Michigan Avenue, Dearborn, Wayne County, Michigan, 48124, United States of America",
  "houseNumber": "22100",
  "postcode": "48124",
  "queryAddr": "22100 Michigan Ave. Dearborn, MI 48124",
  "road": "Michigan Avenue",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Fulton County, Georgia 1`] = `
Object {
  "city": "Atlanta",
  "country": "United States of America",
  "county": "Fulton County",
  "fullAddr": "191 Peachtree Tower, Peachtree Street Northeast, Five Points, Vine City, Atlanta, Fulton County, Georgia, 30303, United States of America",
  "houseNumber": "191 Peachtree Tower",
  "postcode": "30303",
  "queryAddr": "191 Peachtree St NE, Atlanta, GA 30303",
  "road": "Peachtree Street Northeast",
  "state": "Georgia",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Miami-Dade County, Florida 1`] = `
Object {
  "city": "Miami",
  "country": "United States of America",
  "county": "Miami-Dade County",
  "fullAddr": "100, Biscayne Boulevard, The Roads, Miami, Miami-Dade County, Florida, 33131, United States of America",
  "houseNumber": "100",
  "postcode": "33131",
  "queryAddr": "100 S Biscayne Blvd, Miami, FL 33131",
  "road": "Biscayne Boulevard",
  "state": "Florida",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Milwaukee County, Wisconsin 1`] = `
Object {
  "city": "Milwaukee",
  "country": "United States of America",
  "county": "Milwaukee County",
  "fullAddr": "756, North Milwaukee Street, East Town, Milwaukee, Milwaukee County, Wisconsin, 53202, United States of America",
  "houseNumber": "756",
  "postcode": "53202",
  "queryAddr": "756 N Milwaukee St, Milwaukee, WI 53202",
  "road": "North Milwaukee Street",
  "state": "Wisconsin",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Muscogee County, Georgia 1`] = `
Object {
  "city": "Columbus",
  "country": "United States of America",
  "county": "Muscogee County",
  "fullAddr": "1200, 6th Avenue, Rose Hill, Avondale, Columbus, Muscogee County, Georgia, 31901, United States of America",
  "houseNumber": "1200",
  "postcode": "31901",
  "queryAddr": "1200 6th Ave, Columbus, GA 31902",
  "road": "6th Avenue",
  "state": "Georgia",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Muskegon County, Michigan 1`] = `
Object {
  "city": "Muskegon Heights",
  "country": "United States of America",
  "county": "Muskegon County",
  "fullAddr": "2724, Peck Street, Muskegon Heights, Muskegon County, Michigan, 49444, United States of America",
  "houseNumber": "2724",
  "postcode": "49444",
  "queryAddr": "2724 Peck St, Muskegon Heights, MI 49444",
  "road": "Peck Street",
  "state": "Michigan",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Pasco County, Florida 1`] = `
Object {
  "city": "New Port Richey",
  "country": "United States of America",
  "county": "Pasco County",
  "fullAddr": "5443, Main Street, New Port Richey, Pasco County, Florida, 34652, United States of America",
  "houseNumber": "5443",
  "postcode": "34652",
  "queryAddr": "5443 Main St, Port Richey, FL 34652",
  "road": "Main Street",
  "state": "Florida",
  "unit": "1A",
}
`;

exports[`OSM is returning stable results: Richmond County, Georgia 1`] = `
Object {
  "city": "Augusta",
  "country": "United States of America",
  "county": "Richmond County",
  "fullAddr": "Morris Museum of Art, 10th Street, Broad Street Historic District, Augusta, Richmond County, Georgia, 30901, United States of America",
  "houseNumber": "Morris Museum of Art",
  "postcode": "30901",
  "queryAddr": "1 10th St, Augusta, GA 30901",
  "road": "10th Street",
  "state": "Georgia",
  "unit": "1A",
}
`;
